### PR TITLE
[WIP] Ignition Endpoint for Bare-Metal Nodes

### PR DESCRIPTION
**Pull request by BugRoger**
Request a Ignition Userdata via API. Intended use-case is registration of bare-metal nodes that are booted via iPXE

### CODE REVIEW

File: pkg/templates/baremetal_node_1.9.go [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+          --bootstrap-kubeconfig=/etc/kubernetes/bootstrap/kubeconfig \
+          --network-plugin=kubenet \
+          --lock-file=/var/run/lock/kubelet.lock \
+          --exit-on-lock-contention \
+          --pod-manifest-path=/etc/kubernetes/manifests \
+          --allow-privileged \
+          --cluster-dns={{ .ClusterDNSAddress }} \
+          --cluster-domain={{ .ClusterDomain }} \
+          --client-ca-file=/etc/kubernetes/certs/kubelet-clients-ca.pem \
+          --non-masquerade-cidr=0.0.0.0/0 \
```
Comment (databus23) [inline pkg/templates/baremetal_node_1.9.go]: Please add --read-only-port=0. See #264

**Comment by SchwarzM** [comment]
If i read this correct it leaks a lot of information and is currently an authenticated api call.
If you want to use it via ipxe you will either have to think about an authenticated call ( i have done this before it is possible but probably not with actual user credentials ) or an unauthenticated call which relies on the user accepting the csr and providing the needed openstack / cert info later.

**Comment by BugRoger** [comment]
Yeah, this is just a little spike. I want to see what we'd actually have to do to make this work... 

The biggest problem is that the control-plane nodes are all one-off snow-flakes. That just doesn't work very well... We would need to have a mechanism to "register" nodes and their configuration, maybe as a separate `BareMetalNode` CRD.

Alternatively, we could try to reduce the "flakyness" of the nodes. Biggest offender is the network bond config. Which as we already learned is not possible to capture with conventions. 

Would anyone notice that the nodes don't have a bonded interface? Except that it would actually work more reliable? ;)

**Comment by BugRoger** [comment]
Added a new CRD `ExternalNode` that can hold node-specific configuration. The boot IPXE mac address can be used to fish for the correct node.

`curl -K _scratch/.curlrc  localhost:5100/api/v1/clusters/dev/ignition?mac=00:3A:7D:6A:48:F1`

Currently quite simple:

 ` kubectl get externalnode master0.cc.eu-de-1.cloud.sap -o yaml`
```
apiVersion: kubernikus.sap.cc/v1
kind: ExternalNode
metadata:
  name: master0.cc.eu-de-1.cloud.sap
  namespace: kubernikus-michi
spec:
  ipxe: 00:3A:7D:6A:48:F1
```

**Comment by BugRoger** [comment]
Extended the `ExternalNode` to allow for network configuration synonymously to the Bootcfg Ignition Template. Looks like this:

```
apiVersion: kubernikus.sap.cc/v1
kind: ExternalNode
metadata:
  name: master0.cc.bla.cloud.sap

spec:
  ipxe: 12:3A:7D:6A:48:F1
  networks:
    - name: eth0
      match:
        macAddress: 12:df:2f:67:1e:85
      network:
        lldp: "yes"
        dhcp: "no"
        bond: bond1
    - name: eth1
      match:
        macAddress: 70:df:2f:67:1e:86
      network:
        lldp: "yes"
        dhcp: "no"
        bond: bond1
    - name: eth2
      match:
        macAddress: 12:df:2f:66:dc:d5
      network:
        lldp: "yes"
        dhcp: "no"
        bond: bond2
    - name: eth3
      match:
        macAddress: 12:df:2f:66:dc:d6
      network:
        lldp: "yes"
        dhcp: "no"
        bond: bond2
    - name: bond1
      match:
        name: bond1
      network:
        dhcp: "no"
    - name: bond2
      match:
        name: bond2
      network:
        dhcp: "no"
        address: 1.4.7.2/29
        gateway: 1.4.7.1
        dns: ["1.2.9.200", "1.2.9.201"]
        domains: bla..cloud.sap
  netdevs:
    - name: bond1
      netdev:
        name: bond1
        kind: bond
        mtuBytes: 9000
      bond:
        mode: 802.3ad 
        mimMonitorSec: 1s
        lacpTransmitRate: fast
        upDelaySec: 3s
        downDelaySec: 3s
        minLinks: 1
    - name: bond2
      netdev:
        name: bond2
        kind: bond
        mtuBytes: 9000
      bond:
        mode: 802.3ad 
        mimMonitorSec: 1s
        lacpTransmitRate: fast
        upDelaySec: 3s
        downDelaySec: 3s
        minLinks: 1
```

It generates equivalent Ignition. Here the relevant network config:
```
  "networkd": {
    "units": [
      {
        "name": "eth0.network",
        "contents": "[Match]\nMACAddress=12:df:2f:67:1e:85\n[Network]\nDHCP=no\nLLDP=yes\nBond=bond1\n"
      },
      {
        "name": "eth1.network",
        "contents": "[Match]\nMACAddress=12:df:2f:67:1e:86\n[Network]\nDHCP=no\nLLDP=yes\nBond=bond1\n"
      },
      {
        "name": "eth2.network",
        "contents": "[Match]\nMACAddress=12:df:2f:66:dc:d5\n[Network]\nDHCP=no\nLLDP=yes\nBond=bond2\n"
      },
      {
        "name": "eth3.network",
        "contents": "[Match]\nMACAddress=12:df:2f:66:dc:d6\n[Network]\nDHCP=no\nLLDP=yes\nBond=bond2\n"
      },
      {
        "name": "bond1.network",
        "contents": "[Match]\nName=bond1\n[Network]\nDHCP=no\n"
      },
      {
        "name": "bond2.network",
        "contents": "[Match]\nName=bond2\n[Network]\nDHCP=no\nAddress=1.4.7.2/29\nGateway=1.4.7.1\nDNS=1.2.9.200,1.2.9.201\nDomains=kubernetes.bla.cloud.sap\n"
      },
      {
        "name": "bond1.netdev",
        "contents": "[NetDev]\nName=bond1\nKind=bond\nMTUBytes=9000\n[Bond]\nMode=802.3ad\nMIMMonitorSec=1s\nLACPTransmitRate=fast\nUpDelaySec=3s\nDownDelaySec=3s\nMinLinks=1\n"
      },
      {
        "name": "bond2.netdev",
        "contents": "[NetDev]\nName=bond1\nKind=bond\nMTUBytes=9000\n[Bond]\nMode=802.3ad\nMIMMonitorSec=1s\nLACPTransmitRate=fast\nUpDelaySec=3s\nDownDelaySec=3s\nMinLinks=1\n"
      }
    ]
  },
```

In order to retire the whole `region.yaml` config, I would imagine that we create a `NetboxExternalNodeOperator` that fishes in Netbox for the MAC addresses and types of the nodes. It also holds the logic to distinguish the different types of nodes that require slightly different configuration. The Ignition template in Kubernikus can be generic, while the CCloud specifics are outsourced to the  `NetboxExternalNodeOperator`.

File: pkg/apis/kubernikus/v1/externalnode.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+type SystemdNetDevNetDev struct {
+	Name     string `json:"name,omitempty"`
+	Kind     string `json:"kind,omitempty"`
+	MTUBytes int    `json:"mtuBytes,omitempty"`
+}
+
+type SystemdNetDevBond struct {
+	Mode             string `json:"mode,omitempty"`
+	MIMMonitorSec    string `json:"mimMonitorSec,omitempty"`
```
Comment (SchwarzM) [inline pkg/apis/kubernikus/v1/externalnode.go]: This is called MIIMonitor

File: pkg/templates/ignition_test.go [hunk 3]
```diff
@@ -55,7 +55,121 @@ func TestGenerateNode(t *testing.T) {
...
+				kubernikusv1.SystemdNetDevSpec{
+					Name: "bond1",
+					NetDev: &kubernikusv1.SystemdNetDevNetDev{
+						Name:     "bond1",
+						Kind:     "bond",
+						MTUBytes: 9000,
+					},
+					Bond: &kubernikusv1.SystemdNetDevBond{
+						Mode:             "802.3ad",
+						MIMMonitorSec:    "1s",
```
Comment (SchwarzM) [inline pkg/templates/ignition_test.go]: MIIMonitor s.a.

**Pull request closed without merging**